server/models/components/utils.py

```diff
@@ -1,59 +1,75 @@
 from scipy.optimize import fsolve
-
-
-def compute_ratio(dms, max_days):
-    def f(q):
-        return (1 / 2) * (1 - q ** max_days) * (1 + q) / (1 - q) - dms
-
-    solutions = fsolve(f, 5)
-    ratio = solutions[0]
-    return ratio
-
-
-def compute_residuals(dms, max_days):
-    ratio = compute_ratio(dms, max_days)
-    residuals = [ratio ** i for i in range(max_days + 1)]
+import matplotlib.pyplot as plt
+
+
+def factorial(k) :
+    
+    fact = 1
+    
+    for i in range(1, k + 1) :
+        fact *= i
+    
+    return fact
+
+def binomial_coeff(k, n) :
+    
+    return factorial(n) / (factorial(k) * factorial(n - k))
+
+def compute_param(dms, max_days, prob_type):
+    
+    if prob_type == 'exponential':
+        def f(x):
+            return (1 - x ** (max_days)) / (1 - x) - dms
+        solutions = fsolve(f, 5)
+        param = solutions[0]
+    elif prob_type == 'binomial':
+        param = (dms - 1) / max_days
+        
+    return param
+
+def compute_khi(dms, max_days = 21, prob_type = 'binomial') :
+    
+    khi_tab = []
+    param = compute_param(dms, max_days, prob_type)
+    
+    if prob_type == 'exponential':
+        for k in range(max_days - 1):
+            khi = param ** k * (1 - param)
+            khi_tab.append(khi)
+        khi_tab.append(param ** (max_days - 1))
+    elif prob_type == 'binomial':
+        for k in range(max_days):
+            khi = binomial_coeff(k, max_days) * param ** k * (1 - param) ** (max_days - k)
+            khi_tab.append(khi)
+    
+    return khi_tab
+    
+def compute_residuals(khi_tab):
+    
+    residuals = [1]
+    
+    for k in range(len(khi_tab)) :
+        residuals.append(residuals[k] - khi_tab[k])
+    
     return residuals
 
-
-def compute_khi(residuals):
-    max_days = len(residuals) - 1
-    khi = [residuals[i] - residuals[i + 1] for i in range(max_days)]
-    proba_tot = sum(khi)
-
-    for i in range(max_days):
-        khi[i] /= proba_tot
-
-    return khi
-
-
-def compute_exp_ki(duration, max_days=21):
-    return compute_khi(compute_residuals(duration, max_days))
-
-
 def compute_delay_ki(duration):
     if duration <= 1:
         return [1]
     else:
         return [0]*(duration-2) + [0.5, 0.5]
 
 
-def compute_area(dms, max_days):
-
-    residuals = compute_residuals(dms, max_days)
-    khi = compute_khi(residuals)
-
+def compute_area_and_expectation(khi_tab, residuals):
```

Comment:
is this function useful for the library or is it only for testing purpose?
In that case it should be moved in a test file